uefi/src/mem.rs

```diff
@@ -0,0 +1,136 @@
+use core::slice;
+use uefi::data_types::Align;
+use uefi::Error;
+
+/// Helper to return owned versions of certain UEFI data structures on the heap, hence, in a
+/// [`Box`]. It is intended to wrap low-level UEFI functions off this crate that
+/// - consume a mutable reference to a buffer that needs to be filled with some data of type,
+/// - return a mutable typed reference (which points to the same memory as the input buffer) on
+///   success, and
+/// - return the required buffer size on error.
+pub fn make_boxed<
+    'a,
+    Data: Align + ?Sized + Debug + 'a,
+    F: FnMut(&'a mut [u8]) -> Result<&'a mut Data, Option<usize>>,
+>(
+    mut fetch_data_fn: F,
+) -> Result<Box<Data>> {
+    let required_size = match fetch_data_fn(&mut []).map_err(Error::split) {
+        // This is the expected case: the empty buffer passed in is too
+        // small, so we get the required size.
+        Err((Status::BUFFER_TOO_SMALL, Some(required_size))) => Ok(required_size),
+        // Propagate any other error.
+        Err((status, _)) => Err(Error::from(status)),
+        // Success is unexpected, return an error.
+        Ok(_) => Err(Error::from(Status::UNSUPPORTED)),
+    }?;
+
+    // We add trailing padding because the size of a rust structure must
+    // always be a multiple of alignment.
+    let layout = Layout::from_size_align(required_size, Data::alignment())
+        .unwrap()
+        .pad_to_align();
+
+    // Allocate the buffer.
+    let heap_buf: NonNull<u8> = unsafe {
+        let ptr = alloc::alloc(layout);
+        match NonNull::new(ptr) {
+            None => return Err(Status::OUT_OF_RESOURCES.into()),
+            Some(ptr) => ptr,
+        }
+    };
+
+    // Get the file info using the allocated buffer for storage.
+    let data: Result<&mut Data> = {
+        let buffer = unsafe { slice::from_raw_parts_mut(heap_buf.as_ptr(), layout.size()) };
+        fetch_data_fn(buffer).discard_errdata()
+    };
+
+    // If an error occurred, deallocate the memory before returning.
+    let data: &mut Data = match data {
+        Ok(data) => data,
+        Err(err) => {
+            unsafe { alloc::dealloc(heap_buf.as_ptr(), layout) };
+            return Err(err);
+        }
+    };
+
+    let data = unsafe { Box::from_raw(data) };
+
+    Ok(data)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::ResultExt;
+    use core::mem::{align_of, size_of};
+
+    #[derive(Debug)]
+    #[repr(C)]
+    struct SomeData([u8; 4]);
+
+    impl Align for SomeData {
+        fn alignment() -> usize {
+            align_of::<Self>()
+        }
+    }
+
+    /// Function that behaves like the other UEFI functions. It takes a
+    /// mutable reference to a buffer memory that represents a [`SomeData`]
+    /// instance.
+    fn uefi_function_stub_read(buf: &mut [u8]) -> crate::Result<&mut SomeData, Option<usize>> {
+        if buf.len() < 4 {
+            return Status::BUFFER_TOO_SMALL.into_with(|| panic!(), |_| Some(4));
```

Comment:
I think you can use `into_with_err` here to drop the `panic!()`

Reply:
Ah, thanks! much nicer

Reply:
yikes. This is actually hard to compile without the panic trick. I do not find a nice solution. Do you have an idea? @nicholasbishop

Reply:
Does this work?

```rust
return Status::BUFFER_TOO_SMALL.into_with_err(|_| Some(4));
````

Reply:
```
error[E0308]: mismatched types
   --> uefi\src\mem.rs:110:20
    |
105 |     fn uefi_function_stub_read<Data: Align>(buf: &mut [u8]) -> Result<&mut Data, Option<usize>> {
    |                                                                -------------------------------- expected `core::result::Result<&mut Data, result::error::Error<Option<usize>>>` because of return type
...
110 |             return Status::BUFFER_TOO_SMALL.into_with_err(|_| Some(4_usize));
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Data`, found `()`
    |
    = note: expected enum `core::result::Result<&mut Data, _>`
               found enum `core::result::Result<(), _>`

```

Reply:
Ah you are right. I thought I tested it locally but I wasn't compiling the tests so didn't see the error. I think panic!() is probably the right thing to do then.

Reply:
okay, I keep it